source/syntax/CSTSerializer.cpp

```diff
@@ -0,0 +1,202 @@
+//------------------------------------------------------------------------------
+// CSTSerializer.cpp
+// Concrete Syntax Tree JSON serialization
+//
+// SPDX-FileCopyrightText: Michael Popoloski
+// SPDX-License-Identifier: MIT
+//------------------------------------------------------------------------------
+#include "slang/syntax/CSTSerializer.h"
+
+#include "slang/parsing/Token.h"
+#include "slang/syntax/AllSyntax.h"
+#include "slang/syntax/SyntaxTree.h"
+#include "slang/syntax/SyntaxVisitor.h"
+
+// This file contains CST JSON conversion functionality.
+
+namespace slang::syntax {
+
+// Define traits and utility functions for CSTJsonMode
+const std::array<CSTJsonMode, 4> CSTJsonMode_traits::values = {
+    CSTJsonMode::Full, CSTJsonMode::SimpleTrivia, CSTJsonMode::NoTrivia, CSTJsonMode::SimpleTokens};
+
+std::ostream& operator<<(std::ostream& os, CSTJsonMode mode) {
+    return os << toString(mode);
+}
+
+std::string_view toString(CSTJsonMode mode) {
+    switch (mode) {
+        case CSTJsonMode::Full:
+            return "Full";
+        case CSTJsonMode::SimpleTrivia:
+            return "SimpleTrivia";
+        case CSTJsonMode::NoTrivia:
+            return "NoTrivia";
+        case CSTJsonMode::SimpleTokens:
+            return "SimpleTokens";
+    }
+    SLANG_UNREACHABLE;
+}
+
+CSTSerializer::CSTSerializer(JsonWriter& writer, CSTJsonMode mode) : writer(writer), mode(mode) {
+}
+
+void CSTSerializer::serialize(const SyntaxTree& tree) {
+    writer.startObject();
+    writer.writeProperty("kind");
+    writer.writeValue(std::string_view("SyntaxTree"));
+    writer.writeProperty("root");
+    serialize(tree.root());
+    writer.endObject();
+}
+
+struct CSTJsonVisitor : public SyntaxVisitor<CSTJsonVisitor> {
+    JsonWriter& writer;
+    CSTJsonMode mode;
+
+    CSTJsonVisitor(JsonWriter& w, CSTJsonMode m) : writer(w), mode(m) {}
+
+    // Helper methods for common patterns
+    void startSyntaxObject(const SyntaxNode& node) {
+        writer.startObject();
+        writer.writeProperty("kind");
+        writer.writeValue(std::string_view(toString(node.kind)));
+    }
+
+    void writePropertyAndToken(std::string_view name, parsing::Token token) {
+        writer.writeProperty(name);
+        visitToken(token);
+    }
+
+    void writePropertyAndNode(std::string_view name, const SyntaxNode* node) {
+        writer.writeProperty(name);
+        node->visit(*this);
+    }
+
+    void writeOptionalNode(std::string_view name, const SyntaxNode* node) {
+        if (node) {
+            writer.writeProperty(name);
+            node->visit(*this);
+        }
+    }
+
+    void writeOptionalToken(std::string_view name, parsing::Token token) {
+        if (token) {
+            writer.writeProperty(name);
+            visitToken(token);
+        }
+    }
+
+    void writeTokenList(std::string_view name, const TokenList& tokenList) {
+        writer.writeProperty(name);
+        if (tokenList.empty()) {
```

Comment:
If the list is empty should you just skip writing the property altogether? That's what you're doing in the other optional calls above.